cluster/addons/dns/kube2sky/tree.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	pathutil "path"
+	"sort"
+	"strings"
+)
+
+type nodeKind int
+
+const (
+	unset nodeKind = iota
+	file
+	directory
+)
+
+var UnsetNodeErr = errors.New("cannot set etcd from unset node, must be file or directory")
+
+type recordNode struct {
+	name     string
+	data     *string
+	children []*recordNode
+}
+
+func (n *recordNode) kind() nodeKind {
```

Comment:
Given that a node is either a file or a directory and never both, I would expect to see it encoded more as

```
type directory struct {
    name string
    files []*file
    dirs []*directory
}

type file struct {
    name string
    data string
}
```

This has the nice property that the "kind" of a node is not implied by the contents.  Is there any reason to do it this way instead?  I will accept valid arguments in lieu of changes to your code.